Document new experimental ingestion streaming APIs (#8123)
opensearch-project/documentation-website: _api-reference/document-apis/bulk-streaming.md

---
layout: default
title: Streaming bulk
parent: Document APIs
nav_order: 25
redirect_from:
 - /opensearch/rest-api/document-apis/bulk/streaming/
---

# Streaming bulk
**Introduced 2.17.0**
{: .label .label-purple }

> This is an experimental feature and is not recommended for use in a production environment. For updates on the progress of the feature or if you want to leave feedback, see the associated [GitHub issue](https://example.issue.link).    

Reviewer: Please provide the issue # here.

{: .warning}

> The streaming bulk operation lets you add, update, or delete multiple documents in by streaming the request and getting the results as streaming response. In comparison to the traditional [Bulk API]({{site.url}}{{site.baseurl}}/api-reference/document-apis/bulk/), streaming ingestion eliminates the need to guess the batch size (which is affected by the cluster operational state at any given time) and naturally applies the backpressure between many clients and the cluster. The streaming works over HTTP/2 or HTTP/1.1 (using chunked transfer encoding), depending on the capabilities of the clients and the cluster.

Reviewer 2: ```suggestion
The streaming bulk operation lets you add, update, or delete multiple documents by streaming the request and getting the results as a streaming response. In comparison to the traditional [Bulk API]({{site.url}}{{site.baseurl}}/api-reference/document-apis/bulk/), streaming ingestion eliminates the need to estimate the batch size (which is affected by the cluster operational state at any given time) and naturally applies backpressure between many clients and the cluster. The streaming works over HTTP/2 or HTTP/1.1 (using chunked transfer encoding), depending on the capabilities of the clients and the cluster.
```

The default HTTP transport method does not support streaming. You must install the [`transport-reactor-netty4`]({{site.url}}{{site.baseurl}}/install-and-configure/configuring-opensearch/network-settings/#selecting-the-transport) HTTP transport plugin and use it as the default HTTP transport layer. Both the `transport-reactor-netty4` plugin and the Streaming Bulk API are experimental.
{: .note}

## Path and HTTP methods

```json
POST _bulk/stream
POST <index>/_bulk/stream
```

If you specify the index in the path, then you don't need to include it in the [request body chunks]({{site.url}}{{site.baseurl}}/api-reference/document-apis/bulk/#request-body).

> OpenSearch also accepts PUT requests to the `_bulk/stream` path, but we highly recommend using POST. The accepted usage of PUT---adding or replacing a single resource at a given path---doesn't make sense for streaming bulk requests.

Reviewer 2: "on" a given path?

{: .note }

## Query parameters

The following table lists the available query parameters. All query parameters are optional.

Parameter | Data type | Description
:--- | :--- | :---
`pipeline` | String | The pipeline ID for preprocessing documents.

> `refresh` | Enum | Whether to refresh the affected shards after performing the indexing operations. Default is `false`. `true` makes the changes show up in search results immediately, but hurts cluster performance. `wait_for` waits for a refresh. Requests take longer to return, but cluster performance doesn't suffer.

Reviewer 2: ```suggestion
`refresh` | Enum | Whether to refresh the affected shards after performing the indexing operations. Default is `false`. `true` causes the changes show up in search results immediately but degrades cluster performance. `wait_for` waits for a refresh. Requests take longer to return, but cluster performance isn't degraded.
```

Author: So this paragraph is taken as is from existing docs on Bulk API [1], I am fine changing it but it should be done both places. Sounds good to you?

[1] https://opensearch.org/docs/latest/api-reference/document-apis/bulk/#url-parameters

Reviewer: Yes, we can change in both places. I updated the suggestion so it flows nicer.

`require_alias` | Boolean | Set to `true` to require that all actions target an index alias rather than an index. Default is `false`.
`routing` | String | Routes the request to the specified shard.

> `timeout` | Time | How long to wait for the request to return. Default `1m`.

Reviewer 2: ```suggestion
`timeout` | Time | How long to wait for the request to return. Default is `1m`.
```

> `type` | String | (Deprecated) The default document type for documents that don't specify a type. Default is `_doc`. We highly recommend ignoring this parameter and using a type of `_doc` for all indexes.

Reviewer 2: ```suggestion
`type` | String | (Deprecated) The default document type for documents that don't specify a type. Default is `_doc`. We highly recommend ignoring this parameter and using the `_doc` type for all indexes.
```

> `wait_for_active_shards` | String | Specifies the number of active shards that must be available before OpenSearch processes the bulk request. Default is 1 (only the primary shard). Set to `all` or a positive integer. Values greater than 1 require replicas. For example, if you specify a value of 3, the index must have two replicas distributed across two additional nodes for the request to succeed.

Reviewer 2: ```suggestion
`wait_for_active_shards` | String | Specifies the number of active shards that must be available before OpenSearch processes the bulk request. Default is `1` (only the primary shard). Set to `all` or a positive integer. Values greater than 1 require replicas. For example, if you specify a value of 3, then the index must have 2 replicas distributed across 2 additional nodes in order for the request to succeed.
```

> `batch_interval` | Time | Specifies how long bulk operations should be accumulated into batch before sending over to data nodes.

Reviewer 2: ```suggestion
`batch_interval` | Time | Specifies for how long bulk operations should be accumulated into a batch before sending the batch to data nodes.
```

> `batch_size` | Time | Specifies how many bulk operations should be accumulated into batch before sending over to data nodes. Default `1`.

Reviewer 2: ```suggestion
`batch_size` | Time | Specifies how many bulk operations should be accumulated into a batch before sending the batch to data nodes. Default is `1`.
```

> {% comment %}_source | List | asdf

Reviewer 2: Should this and the following two rows be here?

Author: They should (by API specs)

> `_source_excludes` | list | asdf

Reviewer 2: ```suggestion
`_source_excludes` | List | asdf
```

> `_source_includes` | list | asdf{% endcomment %}

Reviewer 2: ```suggestion
`_source_includes` | List | asdf{% endcomment %}
```

## Request body

> The streaming bulk API request body is fully compatible with the [Bulk API request body]({{site.url}}{{site.baseurl}}/api-reference/document-apis/bulk/#request-body), where each bulk operation (create/index/update/delete) is sent as a separate chunk.  

Reviewer 2: ```suggestion
The Streaming Bulk API request body is fully compatible with the [Bulk API request body]({{site.url}}{{site.baseurl}}/api-reference/document-apis/bulk/#request-body), where each bulk operation (create/index/update/delete) is sent as a separate chunk.  
```

## Example request

```json
curl -X POST "http://localhost:9200/_bulk/stream" -H "Transfer-Encoding: chunked" -H "Content-Type: application/json" -d'
{ "delete": { "_index": "movies", "_id": "tt2229499" } }
{ "index": { "_index": "movies", "_id": "tt1979320" } }
{ "title": "Rush", "year": 2013 }
{ "create": { "_index": "movies", "_id": "tt1392214" } }
{ "title": "Prisoners", "year": 2013 }
{ "update": { "_index": "movies", "_id": "tt0816711" } }
{ "doc" : { "title": "World War Z" } }
'
```
{% include copy.html %}

## Example response

> Depending on the batch settings, each streamed response chunk may report the results of one or many (batch) bulk operations, for example for the preceding request with no batching (default), the streaming response may look like this:

Reviewer 2: ```suggestion
Depending on the batch settings, each streamed response chunk may report the results of one or many (batch) bulk operations. For example, for the preceding request with no batching (default), the streaming response may appear as follows:
```
